test/integration/ext/openai/model.py

```diff
@@ -0,0 +1,112 @@
+import pytest
+import vcr
+
+from superduperdb.ext.openai.model import OpenAIChatCompletion, OpenAIEmbedding
+
+
+@vcr.use_cassette(
+    'test/integration/ext/openai/cassettes/test_embed.yaml',
+    filter_headers=['authorization'],
+)
+def test_embed():
+    e = OpenAIEmbedding(model='text-embedding-ada-002')
+    resp = e.predict('Hello, world!')
+
+    assert len(resp) == e.shape[0]
+
+    assert all([isinstance(x, float) for x in resp])
+
+
+@vcr.use_cassette(
+    'test/integration/ext/openai/cassettes/test_batch_embed.yaml',
+    filter_headers=['authorization'],
+)
+def test_batch_embed():
+    e = OpenAIEmbedding(model='text-embedding-ada-002')
+    resp = e.predict(['Hello', 'world!'], batch_size=1)
+
+    assert len(resp) == 2
+
+    assert all([len(x) == e.shape[0] for x in resp])
+
+    assert all([isinstance(x, float) for y in resp for x in y])
+
+
+@vcr.use_cassette(
+    'test/integration/ext/openai/cassettes/test_embed_async.yaml',
+    filter_headers=['authorization'],
+)
+@pytest.mark.asyncio
+async def test_embed_async():
```

Comment:
I just discovered that a lot of async, like FastAPI, is "fake async" where they simply run a separate thread at some point!

No action needed, just noting that.  Well, we need this path because OpenAI offers it, is that the reason?

Reply:
You often don't get the benefit in terms of performance when doing 'fake async' like this. In fact, it can lead to performance degradation sometimes as well. Here isn't the right place to discuss (read: I don't know enough to continue this discussion much further :wink:), but it is part of the reason why I didn't propose that we use an async driver for MongoDB (I think it's called `motor`) because it's fake async and there seems to be questionable performance benefit to it.

From my perspective, I think every framework that makes network calls should offer async functionality, because I think that's what users will expect. I realise that it does increase the maintenance burden though.